src/sparse.c

```diff
@@ -0,0 +1,709 @@
+	*file_exists = git_fs_path_exists(git_str_cstr(&filepath));
+    if (!*file_exists) {
+        if ((error = git_futils_creat_withpath(git_str_cstr(&filepath), 0777, 0666)) < 0)
+			goto done;
+    }
+
+    error = git_attr_cache__get(&sparse->sparse, repo, NULL, &source, parse_sparse_file, false);
+
+done:
+	git_str_dispose(&infopath);
+    return error;
+}
+
+int git_sparse_attr_file__init(
+		git_repository *repo,
+		git_sparse *sparse)
+{
+	int b = false;
+	int error = git_sparse_attr_file__init_(&b, repo, sparse);
+	return error;
+}
+
+int git_sparse__init_(
+		int *file_exists,
+        git_repository *repo,
+        git_sparse *sparse)
+{
+	int error = 0;
+	
+	assert(repo && sparse);
+	
+	memset(sparse, 0, sizeof(*sparse));
+	sparse->repo = repo;
+	
+	/* Read the ignore_case flag */
+	if ((error = git_repository__configmap_lookup(
+			&sparse->ignore_case, repo, GIT_CONFIGMAP_IGNORECASE)) < 0)
+		goto cleanup;
+	
+	if ((error = git_attr_cache__init(repo)) < 0)
+		goto cleanup;
+
+    if ((error = git_sparse_attr_file__init_(file_exists, repo, sparse)) < 0) {
+        if (error != GIT_ENOTFOUND)
+            goto cleanup;
+        error = 0;
+    }
+	
+cleanup:
+	if (error < 0)
+		git_sparse__free(sparse);
+	
+	return error;
+}
+
+int git_sparse__init(
+		git_repository *repo,
+		git_sparse *sparse)
+{
+	int b = false;
+	int error = git_sparse__init_(&b, repo, sparse);
+	return error;
+}
+
+int git_sparse__lookup(
+        int* checkout,
+        git_sparse *sparse,
+        const char* pathname,
+        git_dir_flag dir_flag)
+{
+	git_attr_path path;
+	const char *workdir;
+	int error;
+
+	GIT_ASSERT_ARG(checkout);
+	GIT_ASSERT_ARG(sparse);
+	GIT_ASSERT_ARG(pathname);
+
+	*checkout = GIT_SPARSE_CHECKOUT;
+
+	workdir = git_repository_workdir(sparse->repo);
+	if ((error = git_attr_path__init(&path, pathname, workdir, dir_flag)))
+		return -1;
+	
+	/* No match -> no checkout */
+	*checkout = GIT_SPARSE_NOCHECKOUT;
+	
+	while (1) {
+		if (sparse_lookup_in_rules(checkout, sparse->sparse, &path))
+			goto cleanup;
+		
+		/* move up one directory */
+		if (path.basename == path.path)
+			break;
+		path.basename[-1] = '\0';
+		while (path.basename > path.path && *path.basename != '/')
+			path.basename--;
+		if (path.basename > path.path)
+			path.basename++;
+		path.is_dir = 1;
+	}
+
+cleanup:
+	git_attr_path__free(&path);
+	return 0;
+}
+
+void git_sparse__free(git_sparse *sparse)
+{
+	git_attr_file__free(sparse->sparse);
+}
+
+int git_sparse_checkout__list(
+        git_vector *patterns,
+        git_sparse *sparse)
+{
+    int error = 0;
+    git_str data = GIT_STR_INIT;
+    char *scan, *buf;
+
+    GIT_ASSERT_ARG(patterns);
+    GIT_ASSERT_ARG(sparse);
+
+    if ((error = git_futils_readbuffer(&data, sparse->sparse->entry->fullpath)) < 0)
+        return error;
+
+    scan = (char *)git_str_cstr(&data);
+    while (!error && *scan) {
+
+		buf = git__strtok_(&scan, "\r\n", "\n", "\0");
```

Comment:
This isn't how strtok works. strtok matches any character in the string.
`git__strtok(&scan, "\r\n")` is what you want here.
If either `'\r'` or `'\n'` is encountered it would split on them and empty segments are skipped over.
So "\r\n" would first match on '\r' and then skip over the '\n'.
It would also take "\n\n\r" or any combination of them.

Bit over all it seems a bit strange to have this splitting and parsing here.
https://www.git-scm.com/docs/git-sparse-checkout says:
> By default, the sparse-checkout file uses the same syntax as .gitignore files.

Shouldn't we aim to have the same function for parsing them?

I haven't really thought it through but I have this feeling that it could be possible to merge the spare checkout list and the ignore list for some checks but maybe there is something that prevents that.

Reply:
I've replaced `git__strtok_` with `git__strtok`, I couldn't find any "git ignore list"-like functions around, so I quickly threw this together. I'm open to some ideas on how we can make this list function a bit nicer.

I've also consolidated the ignore parsing functions, so that `parse_sparse_file` now uses `parse_ignore_file` instead. I'd like to keep the `parse_sparse_file` function around since cone mode should change the parsing behaviour a bit.